---
title: 'Bioinformatics in R. Lab 1: dplyr and ggplot'
author: "J. Cesar Ignacio Espinoza - Cesar   "
date: 'Week 01: March 18 and 20, 2024'
output:
  pdf_document: default
  html_document:
    highlight: espresso
    theme: cerulea
---

## Introduction to dplyr
###This is so much fun!
####Yay more data
``

#**dplyr** is a powerful package in R for data manipulation. **It provides** a set of functions that are particularly useful for data wrangling tasks. In this tutorial, we'll cover some of the basic functions in dplyr and how to use them for common data manipulation tasks. Since this is a bioinformatics class the dataset will be RNA-seq FPKM table.

#**Test yourself!!!** Before continuing self test and make sure you can identify the differences between, *FPKM, TPM, RPKM*.
#FPKM, TPM, and RPKM are all measures used in RNA-Seq data analysis to quantify gene expression levels. These measures are crucial in understanding the abundance of transcripts within a sample and comparing gene expression across different samples
#FPKM (Fragments Per Kilobase Million): FPKM is a measure used to quantify gene expression levels in RNA-Seq data. It takes into account the length of the gene and the number of reads that align to it. FPKM normalizes for both the sequencing depth and the gene length, providing a measure of gene expression that can be compared between different genes within a sample.
#TPM (Transcripts Per Million): TPM is another measure used to quantify gene expression levels in RNA-Seq data. Unlike FPKM, TPM normalizes for the total number of reads in a sample and provides a measure of gene expression that can be compared across different samples directly.
#RPKM (Reads Per Kilobase Million): RPKM is similar to FPKM but is used specifically for single-end RNA-Seq data. It calculates the number of reads that align to a gene per kilobase of transcript length per million total reads in the sample.
### Loading the libraries and data. 

We will be using the data from : 	
*Zhang Y, Tong GH, Wei XX, Chen HY et al. Identification of Five Cytotoxicity-Related Genes Involved in the Progression of Triple-Negative Breast Cancer. Front Genet 2021;12:723477.*

```{r message=FALSE, warning=FALSE, paged.print=FALSE, results="hide"}
# We first need to import the important library for today's class, dplyr
library(dplyr)
library(readr)
data <- read_csv('BreastCancerData-Class01.csv')
setwd("C:/Users/nneka/Documents/KGI-2024/Bioinformatics in R/Data Bio in R 2024")
BreastCancerData.Class01 <- read.csv("~/KGI-2024/Bioinformatics in R/Data Bio in R 2024/BreastCancerData-Class01.csv")
View(BreastCancerData.Class01)
data = BreastCancerData.Class01



## Basic dplyr functions

#### Piping

If you took 5120 I hope you remember we covered piping **'|'**, so the output of one function would be passed as the input for another function. here in dplyr piping uses a different operator **'%>%'**

In the code block below pipe the data frame into the base r function to get the names of the columns

```{r}#code box
data%>%
  # Enter the function that will print out the names of the columns. hint there are multiple options
 colnames()
```
```
```{r}
colnames(data)
```

Create more code boxes (formally known as chunks) to inspect the dataset, try the following commands, head(), tail(),  glimpse(), str(), summary(). Write in a sentence or two what they are appear to be doing. You can also look at the documentation.

```{r}
data %>%
  dim()# Returns the dimensions of a matrix or dataframe as a vector with the number of *rows and columns*, providing an overview of the size and shape of the dataset.
```
```{r}
data %>%
  summary() # Generates summary statistics for each variable in a dataframe, including measures of central tendency, dispersion, and quantiles, helping users to understand the distribution of their data
```
```{r}
data%>%
  head()#Returns the first SIX rows of a dataframe or matrix, allowing users to quickly preview the top portion of their data
```
```{r}
data%>%
  tail() # Returns the last SIX rows of a dataframe or matrix, allowing users to quickly preview the bottom portion of their data
```
```{r}
data%>%
  glimpse() #Provides a concise overview of a dataframe, displaying the first few rows along with column names, types, and a sample of the data in each column
```
```{r}
data%>%
  str() # Displays the structure of an R object, providing information on its type, dimensions, and the first few elements or components
```

### Selecting columns. 

The **select()** function is used to select columns from a dataframe. You can specify the columns you want to keep using column names. In the code chunk below select gene, Samples and FPKM, save the subsetted dataframe as subset

```{r}
data
select(data, gene, Samples, FPKM)
subset <- select(data, gene, Samples, FPKM) ### Modify this code block by using the function select
```

**Test yourself!!** What do you think happens when you use a negative sign '-' in front of the column name, Try i.
```{r}
data
select(data, -gene, -Samples, -FPKM)
subset <- select(data, -gene, -Samples, -FPKM) ### The negative sign implies everything but gene, Samples, FPKM
```

### Filtering rows. 

The **filter()** function is used to filter rows based on conditions. You can filter rows from a dataset using logical operators, like =, > or <. If you are comparing strings you need to pass the argument as surrounded by quotation marks. For example : **filter(col_1 == 'string')**
```{r}
data
filter(data, col_1 == 'string')
```
```{r}
metastasis <- data ### Modify this code chunk to generate two dataframes based on the metastasis state of the Sample
no_metastasis <- data 
```

We spent a lot of time in lecture explaing  what RNAseq and FPKM is lets use these data. Subset the dataframe to include only genes with more 200 FPKM as value and are from not cancerous tissue. Get the dimensions of this new dataframe to a new variable called size_not_cancer_abundant and view it. NOTE: not cancer is stored under *characteristics_ch1* and the value associated to not cancer is 	*tissue: normal breast tissue*

```{r}
subset <- data ### Modify this code block
  
```

Select the expression of the following genes BRCA1, BRCA2. and view it

```{r}
brca_exp <- data %>%
  filter(gene == 'BRCA1' | gene == 'BRCA2')
brca_exp
```

### Summarizing Data

The **summarize()** function is used to summarize data, typically by calculating summary statistics. Common operators are mean(), max(), median, sum(), sd(), IQ(). **Try a few!**

```{r}
data %>%
  summarize(average_exp = mean(FPKM))
```

You can pass summarize after another function, remember to use the pipe operator, in the code below calculate the mean FPKM for BRCA1. 

```{r}
data ### pipe your commands here
data ### pipe your commands here

BRCA1_data= data %>%
  filter(gene== "BRCA1")

sum_BRCA1 = BRCA1_data %>%
summarise(average_exp = mean(FPKM), max_exp=max(FPKM))
sum_BRCA1
```

### Grouping data

Another fan favorite from 5120. The **group_by()** function is used to group data by one or more variables. Naturally it is particularly helpful if we use it in combination with other functions. See 

```{r}
data %>%
  filter( gene == 'BRCA1') %>%
  group_by(metastasis.ch1) %>%
  summarize(avg = mean(FPKM))
```
In the code block below, select BRCA1 and BRCA1, calculate the average expression per gene, per tissue per metastasis state. 
```{r}
 ###wRITE YOUR OWN CODE HERE
 data %>%
  filter( gene == 'BRCA1'| gene == "BRCA2") %>%
  group_by(tissue.ch1, gene, metastasis.ch1) %>%
  summarize(avg = mean(FPKM))
```
### Renaming and changing values in a dataframe

To rename columns in a dataframe using dplyr, you can use the **rename()** function. In the code block below, modify rename the columns: metastasis, tissue and characteristics to something more intuitive. Name the new df as renamed and view it.

```{r}
## renamed_df <- df %>%
##    rename(new_name = old_name)
## renamed_df
```

## Introduction to ggplot 

**Test yourself** Before going forward try to see if you remember what gg stands for and what are the three basic components of a plot

In this dataset we will continue exploring the same dataset as thre TNBC as above. First we load the libraries.

```{r message=FALSE, warning=FALSE, paged.print=FALSE, results="hide"}
# We first need to import the libraries 
library(tidyverse)
library(ggplot2)

```

Before diving into plotting lets remember what components of the plot

##### **ggplot(data, aes(x = variable1, y =variable2 )) + geom() + ... + ... +**

#### Bar Plot

```{r}
#  subset <- data %>%
#    filter(gene == 'BRCA1')
#  ggplot(subset, aes(x= subset$Samples, y = subset$FPKM, fill = subset$metastasis.ch1)) + geom_col() 

```
In the example above note how we are using base R to link the specific column using the dolar sign, we can use %>% operator one more time to pass a dataframe as the source of data for our ggplot figures. 

```{R}
data_plot <- data %>%
  filter(gene == 'BRCA1') 
```

We can pass the data frame as the data for our ggplot object


```{r}
#ggplot(data_plot, aes(x=Samples, y = FPKM)) + geom_col() 
### aes(fill = tissue.ch1) 
```
#### Density plots

In the example below pass x= as FPKM, a the aestethics of a ggplot object, base the fill on the tissue type.

```{r}

#ggplot(data_plot, aes(x=FPKM, fill = tissue.ch1)) + geom_density(aes(alpha = 0.5)) 
```

#### Box Plots

```{r}
#data %>%
#  filter(gene == 'BRCA2') %>%
#  ggplot(., aes(x=metastasis.ch1, y = FPKM)) + geom_boxplot()
```

#### Scatter Plots

```{r}
two_genes <- data %>%
  filter(gene == 'BRCA1' | gene == 'BRCA2') %>%
  spread(key = gene, value = FPKM)

#ggplot(two_genes, aes(x=BRCA1, y=BRCA2)) +geom_point() + geom_smooth()

```

#### HeatMap

```{r}
#gene_set <- data %>%
#  filter(gene == 'BRCA1' | gene == 'BRCA2' | gene == 'ALK' | gene == 'MYCN' | gene == 'TP53')

#ggplot(gene_set, aes(x=Samples, y=gene, fill = FPKM)) + geom_tile() 

